jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/ResponseNotifier.java

```diff
@@ -401,9 +408,17 @@ private void onChunk(Content.Chunk chunk)
                 if (LOG.isDebugEnabled())
                     LOG.debug("Registering content in multiplexed content source #{} that contains {}", index, currentChunk);
                 if (currentChunk == null || currentChunk == ALREADY_READ_CHUNK)
-                    this.chunk = chunk.slice();
+                {
+                    Content.Chunk slice = chunk.slice();
+                    // Retain the slice because it is stored for later reads.
+                    if (slice.canRetain())
+                        slice.retain();
```

Comment:
The retain should be done by slice.  If the chunk is not retainable, should it be duplicated here?

Reply:
The retain should not be done by `slice()`, because `slice()` does not know how the slice is used (whether consumed immediately or stored away for later use).
The retain is done here exactly because the sliced chunk is stored for later use, like the comment says.

Reply:
If a chunk is created, then it needs to be released. Methods read, from and slice create a new chunk, so a release should be  needed, so slice needs to retain.   Only if you are passed a chunk as an arg do you not have to release it. Instead you have to retain it if you want to keep it (and then call release).